File: trabalhoFinal/src/trabalho/Funcionario.java

package trabalho;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

/**
 * Representa um funcionário da loja, gerenciando seus dados e histórico de vendas.
 */
public class Funcionario {
    private int idFuncionario;
    private String nome;
    private String cpf;
    private String telefone;
    private String endereco;
    private String cargo;
    private LocalDate dataAdmissao;
    
    private List<Venda> historicoVendas;

    // Construtor
    public Funcionario(int idFuncionario, String nome, String cpf, String telefone, String endereco, String cargo, LocalDate dataAdmissao) {
        this.idFuncionario = idFuncionario;
        this.nome = nome;
        this.cpf = cpf;
        this.telefone = telefone;
        this.endereco = endereco;
        this.cargo = cargo;
        this.dataAdmissao = dataAdmissao;
        
        this.historicoVendas = new ArrayList<>();
    }

    public String getNome() {
        return nome;
    }

    public void registrarVenda(Venda venda) {
        this.historicoVendas.add(venda);
        System.out.println("Venda ID " + venda.getIdVenda() + " registrada para o funcionário " + this.nome);
    }

    public void consultarHistorico() {
        DateTimeFormatter formatador = DateTimeFormatter.ofPattern("dd/MM/yyyy");

        System.out.println("Cargo: " + this.cargo);
        System.out.println("Data de Admissão: " + this.dataAdmissao.format(formatador));
        
        if (this.historicoVendas.isEmpty()) {
            System.out.println("\nNenhuma venda registrada para este funcionário.");
            return; 
        }

        System.out.println("\n-- Vendas Realizadas --");
        double totalVendido = 0.0;
        for (Venda venda : this.historicoVendas) {
            System.out.println(String.format(
                "  - Venda ID: %d | Cliente: %s | Valor: R$%.2f", 
                venda.getIdVenda(), 
                venda.getClienteAssociado().getNome(), 
                venda.getValorTotal()
            ));
            totalVendido += venda.getValorTotal();
        }
        
        System.out.println(String.format("\nValor Total Vendido pelo funcionário: R$%.2f", totalVendido));
        System.out.println("-".repeat(50));
    }
}

File: Python/funcionario.py

from datetime import datetime
from typing import Optional

from pessoa import Pessoa 

class Funcionario(Pessoa):
    
    def __init__(self, nome: str, cpf: str, telefone: str, endereco: str, matricula: str, cargo: str, data_admissao: Optional[datetime] = None):
     
        super().__init__(nome, cpf, telefone, endereco)
        self._matricula = matricula
        self.cargo = cargo
        
        if data_admissao is None:
            self._data_admissao = datetime.now()
        else:
            self._data_admissao = data_admissao
        
        self.historico_vendas = []

    @property
    def matricula(self) -> str:
        return self._matricula

    @property
    def data_admissao(self) -> datetime:
        return self._data_admissao

    def adicionar_venda_ao_historico(self, venda):
        """Este método é chamado pela classe Venda para registrar uma compra."""
        self.historico_vendas.append(venda)

    def consultar_historico(self):
        """Exibe os dados do funcionário e um resumo de seu histórico de vendas."""
        print(f"\n--- Histórico de Vendas do Funcionário: {self.nome} ---")
        
        print(f"Cargo: {self.cargo}")
        print(f"Data de Admissão: {self.data_admissao.strftime('%d/%m/%Y')}")
        
        if not self.historico_vendas:
            print("\nNenhuma venda registrada para este funcionário.")
            return

        print("\n-- Vendas Realizadas --")
        total_vendido = 0
        for venda in self.historico_vendas:
            total_vendido += venda.valor_total
            print(f"  - Venda ID: {venda._id_venda} | Cliente: {venda._cliente.nome} | Valor: R$ {venda.valor_total:.2f}")
        
        print(f"\nValor Total Vendido pelo funcionário: R$ {total_vendido:.2f}")
        print("-" * 50)

    def __str__(self):
        """Representação em string do objeto Funcionário."""
        return f"Funcionário: {self.nome} | Matrícula: {self.matricula} | Cargo: {self.cargo}"